etcdserver/server.go

```diff
@@ -1653,10 +1655,47 @@ func (s *EtcdServer) mayPromoteMember(id types.ID) error {
 	if !s.Cfg.StrictReconfigCheck {
 		return nil
 	}
-	// TODO add more checks whether the member can be promoted.
-	// like learner progress check or if cluster is ready to promote a learner
-	// this is an example to get progress
-	fmt.Printf("raftStatus, %#v\n", raftStatus())
+	if !s.cluster.IsReadyToPromoteMember(uint64(id)) {
+		if lg := s.getLogger(); lg != nil {
+			lg.Warn(
+				"rejecting member promote request; not enough healthy members",
+				zap.String("local-member-id", s.ID().String()),
+				zap.String("requested-member-remove-id", id.String()),
+				zap.Error(ErrNotEnoughStartedMembers),
+			)
+		} else {
+			plog.Warningf("not enough started members, rejecting promote member %s", id)
+		}
+		return ErrNotEnoughStartedMembers
+	}
+
+	// check learner progress is almost catching up the leader.
+	raftStatus := raftStatus()
```

Comment:
Let's put the learner progress check into a separate function. 

`s.Cfg.StrictReconfigCheck == false` (line 1655) should not bypass the progress check. Conceptually `StrictReconfigCheck` is to protect cluster quorum. Learner progress check should not be part of `StrictReconfigCheck`.

Reply:
ok